src/common/src/config.rs

```diff
@@ -35,35 +36,53 @@ pub const STREAM_WINDOW_SIZE: u32 = 32 * 1024 * 1024; // 32 MB
 /// For non-user-facing components where the CLI arguments do not override the config file.
 pub const NO_OVERRIDE: Option<NoOverride> = None;
 
-macro_rules! for_all_config_sections {
-    ($macro:ident) => {
-        $macro! {
-            { server },
-            { meta },
-            { batch },
-            { streaming },
-            { storage },
-            { storage.file_cache },
-        }
-    };
+/// Unrecognized fields in a config section. Generic over the config section type to provide better
+/// error messages.
+///
+/// The current implementation will log warnings if there are unrecognized fields.
+#[derive(Derivative)]
+#[derivative(Clone, Debug, Default)]
+pub struct Unrecognized<T: 'static> {
+    inner: BTreeMap<String, Value>,
+    _marker: std::marker::PhantomData<&'static T>,
 }
 
-macro_rules! impl_warn_unrecognized_fields {
-    ($({ $($field_path:ident).+ },)*) => {
-        fn warn_unrecognized_fields(config: &RwConfig) {
-            if !config.unrecognized.is_empty() {
-                tracing::warn!("unrecognized fields in config: {:?}", config.unrecognized.keys());
-            }
-            $(
-                if !config.$($field_path).+.unrecognized.is_empty() {
-                    tracing::warn!("unrecognized fields in config section [{}]: {:?}", stringify!($($field_path).+), config.$($field_path).+.unrecognized.keys());
-                }
-            )*
-        }
-    };
+impl<T> Unrecognized<T> {
+    /// Returns all unrecognized fields as a map.
+    pub fn into_inner(self) -> BTreeMap<String, Value> {
+        self.inner
+    }
 }
 
-for_all_config_sections!(impl_warn_unrecognized_fields);
+impl<'de, T> Deserialize<'de> for Unrecognized<T> {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let inner = BTreeMap::deserialize(deserializer)?;
+        if !inner.is_empty() {
+            // TODO: may reject unrecognized fields in the future.
+            tracing::warn!(
```

Comment:
I forget why we tolerate these `unrecognized fields` in the first place by not `panicking` it, is it for backward compatibility?

Reply:
I guess so. 🤔 https://github.com/risingwavelabs/risingwave/pull/8325#issue-1608157717